### PR TITLE
Policy Gradient Method using Advantages

### PR DESCRIPTION
**Pull request by luffy1996**
Implementation of Policy gradient methods using Advantages.
Reference : https://karpathy.github.io/2016/05/31/rl/

I am trying to implement policy gradient method using advantages . It will be implemented using category cross-entropy on summation of A(i) * log(p(i)) . Detail implementation of the code can be found in the blog.

I have created policygrad.hpp in reinforcement learning . This header file creates an agent following policy gradient method to learn the game.

I have created advantage error file in methods/layers/ann . This loss function helps in forwards and backward pass.

The policygrad_test.cpp file in methods/test is converging. All together the implementation is complete. Please let me know, if anyone has any suggestion.

### CODE REVIEW

**Comment by luffy1996** [comment]
@zoq , It will be helpful if you go through the codes

**Comment by zoq** [comment]
I'll take a look at the code in the next days.

**Comment by luffy1996** [comment]
Policy is converging. Please look through the codes.

File: src/mlpack/methods/ann/layer/advantage_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Advantage error performance function measures the network's
+ * performance on the basis of advantages.
+ *
+ * Reference : https://karpathy.github.io/2016/05/31/rl/
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/advantage_loss.hpp]: Do you think we can find a paper, that describes the method we can cite here?

File: src/mlpack/methods/ann/layer/advantage_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
+/**
+ * @file advantage_loss.hpp
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/advantage_loss.hpp]: Do you think we should name the file ``advantage_error.hpp`` the same a the class name?

**Comment by zoq** [comment]
@mlpack-jenkins test this please

File: src/mlpack/methods/ann/layer/advantage_loss.hpp (hunk 2 above)
Comment (luffy1996) [reply src/mlpack/methods/ann/layer/advantage_loss.hpp]: I feel that naming it advantage error is to be too specific. It represents class of policy gradient approach , hence I believe we should name it as  policygradient.hpp . This will make it generalised . 

I named it particularly as advantage_error for my luxury to implement the algorithm. However , we can always change it.

File: src/mlpack/methods/ann/layer/advantage_loss.hpp (hunk 1 above)
Comment (luffy1996) [reply src/mlpack/methods/ann/layer/advantage_loss.hpp]: @zoq ,The paper we should cite here is https://arxiv.org/pdf/1506.02438.pdf . The policy gradient method or generalised advantage estimation is what we are trying to solve here . I would like to know your views.

Comment (zoq) [reply src/mlpack/methods/ann/layer/advantage_loss.hpp]: Sounds good.

**Comment by zoq** [comment]
Can you take a look at the style issues?

File: src/mlpack/methods/ann/layer/advantage_loss_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+template<typename InputType, typename TargetType>
+double AdvantageError<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, const TargetType&& target)
+{
+  /*
+  The loss function is mean of log(probability) * target
+  The target captures the advantage of taking the action.
+  The -1 is multiplied to take negative of mean.
+
+  Reference : https://karpathy.github.io/2016/05/31/rl/
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/advantage_loss_impl.hpp]: We should remove the armadillo related documentation here, a user should be able to lookup specific operations.

File: src/mlpack/methods/ann/layer/policygradient.hpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
+/**
+ * @file advantage_loss.hpp
```
Comment (luffy1996) [inline src/mlpack/methods/ann/layer/policygradient.hpp]: @zoq , I changed the name of the file to make it generic. This describes a class of policy gradients (https://arxiv.org/pdf/1506.02438.pdf) (Page 2 :  PRELIMINARIES) .

File: src/mlpack/methods/reinforcement_learning/gae.hpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
+/**
+ * @file policygrad.hpp
```
Comment (luffy1996) [inline src/mlpack/methods/reinforcement_learning/gae.hpp]: @zoq , I have changed this to gae.hpp. Gae stands for generalised advantage estimation. Please provide your inputs

Comment (zoq) [reply src/mlpack/methods/reinforcement_learning/gae.hpp]: Sorry for the slow response on this one, this sounds fine for me.

**Comment by zoq** [comment]
Increasing: https://github.com/mlpack/mlpack/blob/master/src/mlpack/prereqs.hpp#L68 should fix the build issue.

File: src/mlpack/methods/reinforcement_learning/gae.hpp (hunk 5 above)
Comment (luffy1996) [reply src/mlpack/methods/reinforcement_learning/gae.hpp]: Hello @zoq . You might be busy because of GSOC Queries.
However I would like you to have a look of this pull request once you get free.
I have added two of the 6 policy gradient methods as described in the preliminaries of paper.(https://arxiv.org/pdf/1506.02438.pdf) (1 and 2)
 However I am having some difficulties in writing the test case for policy gradients.
You may like to add the codes to the software.

**Comment by luffy1996** [comment]
@zoq The build is successful in my system. Any idea why its failing here.

**Comment by zoq** [comment]
Sorry for the slow response on this one, are you going to open a new PR?

**Comment by luffy1996** [comment]
https://github.com/mlpack/mlpack/pull/1328
Yeah I had some conflicts in my system and I wanted to work with new ideas. For my ease I closed this and opened a new PR. Both the PRs are equivalent.

**Comment by zoq** [comment]
I see, thanks for the clarification.
